graql/reasoner/query/DisjunctiveQuery.java

```diff
@@ -0,0 +1,304 @@
+        return new DisjunctiveQuery(
+                getClauses().stream().map(q -> q.withSubstitution(sub)).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    public Stream<ConceptMap> resolve(Set<ReasonerAtomicQuery> subGoals, boolean infer) {
+
+        boolean doNotResolve = !infer || getAtoms().isEmpty() || (isPositive() && !isRuleResolvable());
+        if (doNotResolve) {
+            Stream<Stream<ConceptMap>> answerStreams = clauses.stream().map(clause ->
+                    traversalExecutor.traverse(clause.getPattern()).map(ans -> {
+                        ConceptMap clauseAns = new ConceptMap(ans.map(), new LookupExplanation(), clause.getPattern(ans.map()));
+                        HashMap<Variable, Concept> bindingVarsSub = filterBindingVars(ans.map());
+                        return new ConceptMap(bindingVarsSub, new DisjunctiveExplanation(clauseAns), getPattern(bindingVarsSub));
+                    }));
+            LazyMergingStream<ConceptMap> mergedStreams = new LazyMergingStream<>(answerStreams);
+            return mergedStreams.flatStream();
+        } else {
+            return new ResolutionIterator(this, subGoals, context().queryCache()).hasStream();
+        }
+    }
+
+    @Override
+    ResolvableQuery inferTypes() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(CompositeQuery::inferTypes).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery constantValuePredicateQuery() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(CompositeQuery::constantValuePredicateQuery).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery copy() {
+        return new DisjunctiveQuery(
+                new HashSet<>(getClauses()),
+                getBindingVars(),
+                traversalExecutor,
+                context());
+    }
+
+    @Override
+    public boolean isAtomic() {
+        // TODO unclear of the meaning of atomicity in this case
+        return getClauses().stream().allMatch(CompositeQuery::isAtomic);
+    }
+
+    @Override
+    boolean isEquivalent(ResolvableQuery q) {
+        DisjunctiveQuery that = q.asDisjunctive();
+        return that.getClauses().size() == getClauses().size()
+                && getClauses().stream().allMatch(clause -> that.getClauses().stream().anyMatch(clause::isEquivalent));
+    }
+
+    @Override
+    public String toString(){
+        return getPattern().toString();
+    }
+
+    @Override
+    public ReasonerQuery conjunction(ReasonerQuery q) {
+        throw new UnsupportedOperationException();
```

Comment:
Delete as the method is moved out to be a helper

Reply:
Actually this fits well with the hierarchy and as a helper may be an anti-pattern of `instanceOf`checks against each type